### PR TITLE
CFE-3738: Clarified when and where custom promises are evaluated

### PR DESCRIPTION
**Pull request by olehermanse**

### CODE REVIEW

File: reference/promise-types/custom.markdown [hunk 1]
```diff
@@ -95,6 +96,13 @@ Due to the implementation details, the following attributes from the `classes` b
 * `persist_time`
 * `timer_policy`
 
+### Evaluation passes and normal order
+
+In CFEngine, each bundle is evaluated in multiple passes (3 main passes for most promise types).
+Within each evaluation pass of a bundle, the promises are not evaluated from top to bottom, but based on a [normal order][Normal Ordering] of the promise types.
+Custom promise types are added dynamically and don't have a predefined order, they are evaluated as they appear within a bundle (top to bottom), but at the end of each evaluation pass, after all the built in promise types.
```
Comment (nickanderson) [inline reference/promise-types/custom.markdown:103]: Evaluated as they appear?

So, this policy file http promise would happen after reports and before git:

```
http:
    "https://cfengine.com/images/cfengine-logo.svg"
      file => "/var/cfengine/cfengine-logo.svg",
      if => not(fileexists("/var/cfengine/cfengine-logo.svg"));
  git:
    "starter_pack_repo"
      repository => "https://github.com/cfengine/starter_pack",
      destination => "/northern.tech/cfengine/starter-pack",
      version => "master";
```

And in another policy file, git would happen after reports before http:

```
git:
    "starter_pack_repo"
      repository => "https://github.com/cfengine/starter_pack",
      destination => "/northern.tech/cfengine/starter-pack",
      version => "master";
http:
    "https://cfengine.com/images/cfengine-logo.svg"
      file => "/var/cfengine/cfengine-logo.svg",
      if => not(fileexists("/var/cfengine/cfengine-logo.svg"));
```

Does that also apply to those that are interleaved? Meaning that after reports the git promise for starter pack would get actuated, and then the http promise type and then the git promise type for documentation? 

```
git:
    "starter_pack_repo"
      repository => "https://github.com/cfengine/starter_pack",
      destination => "/northern.tech/cfengine/starter-pack",
      version => "master";

http:
    "https://cfengine.com/images/cfengine-logo.svg"
      file => "/var/cfengine/cfengine-logo.svg",
      if => not(fileexists("/var/cfengine/cfengine-logo.svg"));

git:
    "cfengine-documentation"
      repository => "https://github.com/cfengine/documentation",
      destination => "/northern.tech/cfengine/documentation",
      version => "master";
```

**Review by nickanderson — CHANGES_REQUESTED** [review]
Can you add a bit about how it applies in the case of interleaving custom promise types?

**Comment by nickanderson** [comment]
@cf-bottom jenkins, please

File: reference/promise-types/custom.markdown (hunk 1 above)
Comment (olehermanse) [reply reference/promise-types/custom.markdown:103]: The _promise types_ are evaluated in the order they appear, not the individual promises.

Just like with built in promise types, adding repeating sections of the same promise types or even interleaving them has no effect, it's the same as putting all the promises with the same type in one section.

I think we should change it, it makes more sense to me if all the promises were just evaluated in order, but at least it's somewhat consistent now.

**Comment by nickanderson** [comment]
@cf-bottom jenkins, please

**Comment by cf-bottom** [comment]
Sure, I triggered a build:

[Build Status](https://ci.cfengine.com//job/fast-build-and-deploy-docs-master/233/)

**Jenkins:** https://ci.cfengine.com/job/fast-build-and-deploy-docs-master/233/

**Documentation:** http://buildcache.cfengine.com/packages/build-documentation-pr/jenkins-fast-build-and-deploy-docs-master-233/output/_site/

File: reference/promise-types/custom.markdown [hunk 2]
```diff
@@ -95,6 +96,13 @@ Due to the implementation details, the following attributes from the `classes` b
...
+As with other promise types, we recommend not relying too much on this ordering, if you want some promises to be evaluated before others, use the `bundlesequence` or `depends_on` attribute to achieve this.
```
Comment (olehermanse) [inline reference/promise-types/custom.markdown:104]:
```suggestion
As with other promise types, we recommend not relying too much on this ordering, if you want some promises to be evaluated before others, use the `bundlesequence` or `depends_on` attribute to achieve this.

**Note:** All promises of the same type are evaluated together, so splitting up the promises of one type or interleaving promises of multiple types will not make a difference. All promises of the custom promise type which appeared first will be evaluated before all the promises of the custom promise type which appeared second are evaluated, and so on.
```

**Comment by olehermanse** [comment]
@cf-bottom jenkins, please

**Comment by cf-bottom** [comment]
Sure, I triggered a build:

[Build Status](https://ci.cfengine.com//job/fast-build-and-deploy-docs-master/234/)

**Jenkins:** https://ci.cfengine.com/job/fast-build-and-deploy-docs-master/234/

**Documentation:** http://buildcache.cfengine.com/packages/build-documentation-pr/jenkins-fast-build-and-deploy-docs-master-234/output/_site/

**Review by craigcomstock — APPROVED** [review]
Yes. I like this clarification. 🎉

**Comment by olehermanse** [comment]
Cherry-pick: https://github.com/cfengine/documentation/pull/2914

File: reference/promise-types/custom.markdown (hunk 1 above)
Comment (nickanderson) [reply reference/promise-types/custom.markdown:103]: OK, and it's about their USE within a bundle, **not** the order in which they are defined (order of promise agent blocks)

**Pull request merged**